src/hotspot/os/linux/os_linux.cpp

```diff
@@ -3470,8 +3480,23 @@ bool os::remove_stack_guard_pages(char* addr, size_t size) {
 // may not start from the requested address. Unlike Linux mmap(), this
 // function returns null to indicate failure.
 static char* anon_mmap(char* requested_addr, size_t bytes) {
-  // MAP_FIXED is intentionally left out, to leave existing mappings intact.
-  const int flags = MAP_PRIVATE | MAP_NORESERVE | MAP_ANONYMOUS;
+  // If a requested address was given:
+  //
+  // The POSIX-conform way is to *omit* MAP_FIXED. This will leave existing mappings intact.
+  // If the requested mapping area is blocked by a pre-existing mapping, the kernel will map
+  // somewhere else. On Linux, that alternative address appears to have no relation to the
+  // requested address.
+  // Unfortunately, this is not what we need - if we specified a request address, we'd want
+  // to map there and nowhere else. Therefore we will unmap the block again, which means we
+  // just executed a needless mmap->munmap cycle.
+  // Since Linux 4.17, the kernel offers MAP_FIXED_NOREPLACE. With this flag, if a pre-
+  // existing mapping exists, the kernel will not map at an alternative point but instead
+  // return an error. We can therefore save that unnessassary mmap-munmap cycle.
+  //
+  // Backward compatibility: Older kernels will ignore the unknown flag; so mmap will behave
+  // as in mode (a).
```

Comment:
Is this in case of a libc that's more up to date than the kernel?

Reply:
The libc has little to do with it. mmap is a system call, and the flags should come from mman.h, which should be a kernel header. 

But I am not sure whether the flags we use when including mman.h are actually directly from the kernel, because the glibc contains the same flags and talks about "kernel header is not namespace clean", whatever that means.

Bottomline, on the build machine we either have the flag - from libc or kernel - or we don't, in which case we define it ourselves. At runtime, the kernel either understands the flag or it don't, in the latter case it shears the flag off and then behaves as if the flag had not been given.

See also man 2 mmap:

```
Note  that older kernels which do not recognize the MAP_FIXED_NOREPLACE flag will typically (upon detecting a collision with a preexisting mapping) fall back to a "non-MAP_FIXED" type of behavior: they will return an address that is different from the requested address.  Therefore, backward-compatible software should check the returned address against the requested address.
```